### PR TITLE
[v1.6.x] Backport [MKL-DNN] Integrate Conv3d and Pool3d/1d (#17884) and Fix Sanity pipeline in 1.6.x

### PR DESCRIPTION
**Pull request by ChaiBapchya**
* Integrate MKl-DNN conv3d and pool3d/1d

* fix UT & address comments

* clean code

* rebase against latest master

### CODE REVIEW

**Comment by ChaiBapchya** [comment]
@mxnet-bot run ci [sanity]

**Comment by TaoLv** [comment]
@wuxun-zhang please help to review ~

File: src/operator/nn/mkldnn/mkldnn_pooling-inl.h [hunk 1]
```diff
@@ -90,22 +88,47 @@ class MKLDNNPoolingBwd {
...
-                                 const mxnet::TShape &dshape) {
-  bool ret = SupportMKLDNNPooling(param);
-  if (!ret)
+                                 const NDArray &input) {
+  const auto dshape = input.shape();
+  const auto ndim = dshape.ndim();
+  const auto dtype = input.dtype();
+
+  if (!(SupportStorageMKLDNN(input.storage_type()) && (ndim == 3 || ndim == 4 || ndim == 5) &&
+       (dtype == mshadow::kFloat32 || dtype == mshadow::kBfloat16)))
```
Comment (wuxun-zhang) [inline src/operator/nn/mkldnn/mkldnn_pooling-inl.h:107]: same as before. Please remove `dtype == mshadow::kBfloat16`.

File: src/operator/subgraph/mkldnn/mkldnn_subgraph_base-inl.h [hunk 2]
```diff
@@ -29,7 +29,9 @@ static inline bool SupportMKLDNNAttr(const std::shared_ptr<NodeAttr>& node_attr)
   if (node_attr) {
     int ndim = node_attr->ishape[0].ndim();
     return (node_attr->dispatch_mode == DispatchMode::kFComputeEx) &&
-           (node_attr->itype[0] == mshadow::kFloat32) && (ndim == 1 || ndim == 2 || ndim == 4);
+           (node_attr->itype[0] == mshadow::kFloat32 ||
+            node_attr->itype[0] == mshadow::kBfloat16) &&
```
Comment (wuxun-zhang) [inline src/operator/subgraph/mkldnn/mkldnn_subgraph_base-inl.h:33]: same as before. Please remove `dtype == mshadow::kBfloat16`.

File: src/operator/nn/mkldnn/mkldnn_base-inl.h [hunk 3]
```diff
@@ -129,15 +129,8 @@ static inline bool SupportMKLDNN(int dtype, const mxnet::TShape &shape) {
     // MKLDNN currently does not support 0-dim Tensor and 0-size Tensor
     return false;
   }
-  return dtype == mshadow::kFloat32 && (ndim == 1 || ndim == 2 || ndim == 4);
-}
-
-static inline bool SupportMKLDNNRnn(const NDArray &input) {
```
[the remark this answers is not in the document]
Comment (zixuanweeei) [reply src/operator/nn/mkldnn/mkldnn_base-inl.h:135]: I think we can drop this function.

File: src/operator/nn/mkldnn/mkldnn_act.cc [hunk 4]
```diff
@@ -48,10 +48,10 @@ bool SupportMKLDNNAct(const ActivationParam& param) {
 }
 
 bool SupportMKLDNNAct(const ActivationParam& param, const NDArray &input) {
-  // MKL-DNN Activation supports 1d, 2d, 3d, 4d data layout
+  // MKL-DNN Activation supports 1d, 2d, 3d, 4d and 5d data layout
   if ((input.shape().ndim() < 1) ||
-      (input.shape().ndim() > 4) ||
-      (input.dtype() != mshadow::kFloat32))
+      (input.shape().ndim() > 5) ||
+      !(input.dtype() == mshadow::kFloat32 || input.dtype() == mshadow::kBfloat16))
```
[the remark this answers is not in the document]
Comment (ChaiBapchya) [reply src/operator/nn/mkldnn/mkldnn_act.cc:54]: Addressed.

File: src/operator/nn/mkldnn/mkldnn_act.cc [hunk 5]
```diff
@@ -62,10 +62,10 @@ bool SupportMKLDNNLeakyRelu(const LeakyReLUParam& param) {
 }
 
 bool SupportMKLDNNLeakyRelu(const LeakyReLUParam& param, const NDArray &input) {
-  // MKL-DNN Activation supports 1d, 2d, 3d, 4d data layout
+  // MKL-DNN Activation supports 1d, 2d, 3d, 4d and 5d data layout
   if ((input.shape().ndim() < 1) ||
-      (input.shape().ndim() > 4) ||
-      (input.dtype() != mshadow::kFloat32))
+      (input.shape().ndim() > 5) ||
+      !(input.dtype() == mshadow::kFloat32 || input.dtype() == mshadow::kBfloat16))
```
[the remark this answers is not in the document]
Comment (ChaiBapchya) [reply src/operator/nn/mkldnn/mkldnn_act.cc]: Addressed

File: src/operator/nn/mkldnn/mkldnn_base-inl.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChaiBapchya) [reply src/operator/nn/mkldnn/mkldnn_base-inl.h:135]: Thanks for confirming!

File: src/operator/nn/mkldnn/mkldnn_base-inl.h [hunk 6]
```diff
@@ -129,15 +129,8 @@ static inline bool SupportMKLDNN(int dtype, const mxnet::TShape &shape) {
...
-  if (input.dtype() == mshadow::kFloat32 && input.shape().ndim() == 3
-      && dmlc::GetEnv("MXNET_USE_MKLDNN_RNN", 1)) {
-    return true;
-  }
-  return false;
+  return (dtype == mshadow::kFloat32 || dtype == mshadow::kBfloat16) &&
```
[the remark this answers is not in the document]
Comment (ChaiBapchya) [reply src/operator/nn/mkldnn/mkldnn_base-inl.h]: Addressed

File: src/operator/nn/mkldnn/mkldnn_pooling-inl.h (hunk 1 above)
Comment (ChaiBapchya) [reply src/operator/nn/mkldnn/mkldnn_pooling-inl.h]: Addressed

File: src/operator/subgraph/mkldnn/mkldnn_subgraph_base-inl.h (hunk 2 above)
Comment (ChaiBapchya) [reply src/operator/subgraph/mkldnn/mkldnn_subgraph_base-inl.h]: Addressed

**Comment by ChaiBapchya** [comment]
@wuxun-zhang Thanks for pointing it out. Addressed your comments.

**Comment by ChaiBapchya** [comment]
`test_quantization_mkldnn.test_quantized_conv` failed with
```
[2020-05-04T03:07:42.847Z] ----------------------------------------------------------------------

[2020-05-04T03:07:42.847Z] Traceback (most recent call last):

[2020-05-04T03:07:42.847Z]   File "/usr/local/lib/python3.5/dist-packages/nose/case.py", line 198, in runTest

[2020-05-04T03:07:42.847Z]     self.test(*self.arg)

[2020-05-04T03:07:42.847Z]   File "/usr/local/lib/python3.5/dist-packages/nose/util.py", line 620, in newfunc

[2020-05-04T03:07:42.847Z]     return func(*arg, **kw)

[2020-05-04T03:07:42.847Z]   File "/work/mxnet/tests/python/mkl/../unittest/common.py", line 177, in test_new

[2020-05-04T03:07:42.847Z]     orig_test(*args, **kwargs)

[2020-05-04T03:07:42.847Z]   File "/work/mxnet/tests/python/mkl/../quantization/test_quantization.py", line 283, in test_quantized_conv

[2020-05-04T03:07:42.847Z]     check_quantized_conv((1, 3, 4, 28, 28), (1, 3, 3), 128, (1, 1, 1), (1, 1, 1), (2, 2, 2), False, qdtype)

[2020-05-04T03:07:42.847Z]   File "/work/mxnet/tests/python/mkl/../quantization/test_quantization.py", line 217, in check_quantized_conv

[2020-05-04T03:07:42.847Z]     arg_shapes, _, _ = conv.infer_shape(data=data_shape)

[2020-05-04T03:07:42.847Z]   File "/work/mxnet/python/mxnet/symbol/symbol.py", line 1103, in infer_shape

[2020-05-04T03:07:42.847Z]     res = self._infer_shape_impl(False, *args, **kwargs)

[2020-05-04T03:07:42.847Z]   File "/work/mxnet/python/mxnet/symbol/symbol.py", line 1267, in _infer_shape_impl

[2020-05-04T03:07:42.847Z]     ctypes.byref(complete)))

[2020-05-04T03:07:42.847Z]   File "/work/mxnet/python/mxnet/base.py", line 255, in check_call

[2020-05-04T03:07:42.847Z]     raise MXNetError(py_str(_LIB.MXGetLastError()))

[2020-05-04T03:07:42.847Z] mxnet.base.MXNetError: Error in operator conv: [03:07:30] src/operator/nn/convolution.cc:242: Check failed: param_.dilate.Size() == 1U (8 vs. 1) : Dilate is not supported in 3d convolution
```

Cherrypicked #17491 which removed the conv3d dilate restriction.
That should resolve unix-cpu pipeline errors.

**Comment by ChaiBapchya** [comment]
@wuxun-zhang One last failure
unix-gpu pipeline
Tests / Cpp: MKLDNN+GPU
```
pooling-inl.h:313: Check failed: outputs.size() == GetNumOutputs(param) (1 vs. 2) :  
```
I was able to repro locally. Not sure why output size is mismatch for pooling.

**Comment by wuxun-zhang** [comment]
@ChaiBapchya To fix the unix-gpu issue, you can refer to [this link](https://github.com/apache/incubator-mxnet/pull/18064/commits/ea1c0807b54aa2667b3b08d116a2844987705450), like what I did when backported this PR to v1.x branch before.

**Comment by ChaiBapchya** [comment]
@mxnet-bot run ci [unix-cpu, unix-gpu]

**Comment by ChaiBapchya** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by ChaiBapchya** [comment]
@mxnet-label-bot add [pr-awaiting-merge]

**Pull request merged**